File: vs2013_test/test_mjlib/tbl/table_mgr.cpp
#include "stdafx.h"
#include <stdio.h>
#include "table_mgr.h"
#include "set_table.h"
#include "hash_table.h"
#include "array_table.h"
#include "binary_array_table.h"

#define TABLECLASS BinaryArrayTable

TableMgr::TableMgr()
{
    for(int i=0; i<9; ++i)
    {
		m_check_table[i] = new TABLECLASS;
    }

    for(int i=0; i<9; ++i)
    {
		m_check_eye_table[i] = new TABLECLASS;
    }
 
    for(int i=0; i<9; ++i)
    {
		m_check_feng_table[i] = new TABLECLASS;
    }

    for(int i=0; i<9; ++i)
    {
		m_check_feng_eye_table[i] = new TABLECLASS;
    }
}

TableMgr::~TableMgr()
{
    for(int i=0; i<9; ++i)
    {
        delete m_check_table[i];
    }

    for(int i=0; i<9; ++i)
    {
        delete m_check_eye_table[i];
    }
 
    for(int i=0; i<9; ++i)
    {
        delete m_check_feng_table[i];
    }

    for(int i=0; i<9; ++i)
    {
        delete m_check_feng_eye_table[i];
    }
}

TableMgr* TableMgr::get_instance()
{
    static TableMgr mgr;
    return &mgr;
}

bool TableMgr::check(int key, int gui_num, bool eye, bool chi)
{
    Table* tbl = 0;

    if(chi)
    {
        if(eye)
        {
            tbl = m_check_eye_table[gui_num];
        }
        else
        {
            tbl = m_check_table[gui_num];
        }
    }
    else
    {
        if(eye)
        {
            tbl = m_check_feng_eye_table[gui_num];
        }
        else
        {
            tbl = m_check_feng_table[gui_num];
        }
    }

    return tbl->check(key);
}

void TableMgr::add(int key, int gui_num, bool eye, bool chi)
{
    Table* tbl = 0;

    if(chi)
    {
        if(eye)
        {
            tbl = m_check_eye_table[gui_num];
        }
        else
        {
            tbl = m_check_table[gui_num];
        }
    }
    else
    {
        if(eye)
        {
            tbl = m_check_feng_eye_table[gui_num];
        }
        else
        {
            tbl = m_check_feng_table[gui_num];
        }
    }

    tbl->add(key);
}

bool TableMgr::load()
{
    char path[256];
    for(int i=0; i<9; ++i)
    {
        sprintf(path, "tbl/table_%d.tbl", i);
        m_check_table[i]->load(path);
    }

    for(int i=0; i<9; ++i)
    {
        sprintf(path, "tbl/eye_table_%d.tbl", i);
        m_check_eye_table[i]->load(path);
    }
 
    for(int i=0; i<9; ++i)
    {
        sprintf(path, "tbl/feng_table_%d.tbl", i);
        m_check_feng_table[i]->load(path);
    }

    for(int i=0; i<9; ++i)
    {
        sprintf(path, "tbl/feng_eye_table_%d.tbl", i);
        m_check_feng_eye_table[i]->load(path);
    }
    return true;
}

bool TableMgr::dump_table()
{
    char path[256];
    for(int i=0; i<9; ++i)
    {
        sprintf(path, "tbl/table_%d.tbl", i);
        m_check_table[i]->dump(path);
    }

    for(int i=0; i<9; ++i)
    {
        sprintf(path, "tbl/eye_table_%d.tbl", i);
        m_check_eye_table[i]->dump(path);
    }
    return true;
}

bool TableMgr::dump_feng_table()
{
    char path[256];
    for(int i=0; i<9; ++i)
    {
        sprintf(path, "tbl/feng_table_%d.tbl", i);
        m_check_feng_table[i]->dump(path);
    }

    for(int i=0; i<9; ++i)
    {
        sprintf(path, "tbl/feng_eye_table_%d.tbl", i);
        m_check_feng_eye_table[i]->dump(path);
    }

    return true;
}

File: mjlib_java/table/src/org/yungege/mj/TableMgr.java
package org.yungege.mj;

public class TableMgr {
	public static TableMgr mgr = new TableMgr();
	public SetTable[] m_check_table = new SetTable[9];
	public SetTable[] m_check_eye_table = new SetTable[9];
	public SetTable[] m_check_feng_table = new SetTable[9];
	public SetTable[] m_check_feng_eye_table = new SetTable[9];

	public TableMgr() {
		for (int i = 0; i < 9; ++i) {
			m_check_table[i] = new SetTable();
			m_check_eye_table[i] = new SetTable();
			m_check_feng_table[i] = new SetTable();
			m_check_feng_eye_table[i] = new SetTable();
		}
	}

	public static TableMgr getInstance() {
		return mgr;
	}

	public boolean check(int key, int gui_num, boolean eye, boolean chi) {
		SetTable tbl;

		if (chi) {
			if (eye) {
				tbl = m_check_eye_table[gui_num];
			} else {
				tbl = m_check_table[gui_num];
			}
		} else {
			if (eye) {
				tbl = m_check_feng_eye_table[gui_num];
			} else {
				tbl = m_check_feng_table[gui_num];
			}
		}

		return tbl.check(key);
	}

	public void add(int key, int gui_num, boolean eye, boolean chi) {
		SetTable tbl;

		if (chi) {
			if (eye) {
				tbl = m_check_eye_table[gui_num];
			} else {
				tbl = m_check_table[gui_num];
			}
		} else {
			if (eye) {
				tbl = m_check_feng_eye_table[gui_num];
			} else {
				tbl = m_check_feng_table[gui_num];
			}
		}

		tbl.add(key);
	}

	public boolean load() {
		for (int i = 0; i < 9; ++i) {
			String path = "tbl/table_";
			m_check_table[i].load(path + i + ".tbl");
		}

		for (int i = 0; i < 9; ++i) {
			String path = "tbl/eye_table_";
			m_check_eye_table[i].load(path + i + ".tbl");
		}

		for (int i = 0; i < 9; ++i) {
			String path = "tbl/feng_table_";
			m_check_feng_table[i].load(path + i + ".tbl");
		}

		for (int i = 0; i < 9; ++i) {
			String path = "tbl/feng_eye_table_";
			m_check_feng_eye_table[i].load(path + i + ".tbl");
		}
		return true;
	}

	public boolean dump_table() {
		for (int i = 0; i < 9; ++i) {
			String path = "tbl/table_";
			m_check_table[i].dump(path + i + ".tbl");
		}

		for (int i = 0; i < 9; ++i) {
			String path = "tbl/eye_table_";
			m_check_eye_table[i].dump(path + i + ".tbl");
		}
		return true;
	}

	public boolean dump_feng_table() {
		for (int i = 0; i < 9; ++i) {
			String path = "tbl/feng_table_";
			m_check_feng_table[i].dump(path + i + ".tbl");
		}

		for (int i = 0; i < 9; ++i) {
			String path = "tbl/feng_eye_table_";
			m_check_feng_eye_table[i].dump(path + i + ".tbl");
		}

		return true;
	}
}
